tools/rpkg/tests/testthat/test_relational.R

```diff
@@ -314,6 +312,234 @@ test_that("Symmetric difference returns the symmetric difference", {
+    order_over_window <- rapi_rel_order(proj_window, list(expr_reference("r")))
+    expected_result <- data.frame(r=c(1:4), str_agg_res=c("hello,Big", "hello,Big", "world,42", "world,42"))
+    res <- rel_to_altrep(order_over_window)
+    expect_equal(res, expected_result)
+})
+
+test_that("You can perform window functions on row_number", {
+	  # select a, b, row_number() OVER () from tmp order by a;
+    rel_a <- rel_from_df(con, data.frame(a=c(8:1),b=c(1, 1, 2, 2, 3, 3, 4, 4)))
+    row_number <- expr_function("row_number", list())
+    window_function <- expr_window(row_number)
+    expr_set_alias(window_function, "row_number")
+    proj <- rel_project(rel_a, list(expr_reference("a"), window_function))
+    order_by_a <- rel_order(proj, list(expr_reference("a")))
+    expected_result <- data.frame(a=c(1:8), row_number=(8:1))
+    res <- rel_to_altrep(order_by_a)
+    expect_equal(res, expected_result)
+})
+
+# also tests order by inside the rank function.
+# these tests come from https://dplyr.tidyverse.org/articles/window-functions.html
+# in dplyr min_rank = rank
+test_that("You can perform the window function min_rank", {
+    rel_a <- rel_from_df(con, data.frame(a=c(1, 1, 2, 2, 2)))
+    rank_func <- expr_function("rank", list(), order_bys=list(expr_reference("a")))
+    min_rank_window <- expr_window(rank_func)
+    expr_set_alias(min_rank_window, "window_result")
+	  window_proj <- rel_project(rel_a, list(expr_reference("a"), min_rank_window))
+	  res <- rel_to_altrep(window_proj)
+	  expected_result <- data.frame(a=c(1, 1, 2, 2, 2), window_result=c(1, 1, 3, 3, 3))
+    expect_equal(res, expected_result)
+})
+
+test_that("You can perform the window function dense_rank", {
+    rel_a <- rel_from_df(con, data.frame(a=c(1, 1, 2, 2, 2)))
+    dense_rank_fun <- expr_function("dense_rank", list(), order_bys=list(expr_reference("a")))
+    min_rank_window <- expr_window(dense_rank_fun)
+    expr_set_alias(min_rank_window, "window_result")
+	  window_proj <- rel_project(rel_a, list(expr_reference("a"), min_rank_window))
+	  res <- rel_to_altrep(window_proj)
+	  expected_result <- data.frame(a=c(1, 1, 2, 2, 2), window_result=c(1, 1, 2, 2, 2))
+    expect_equal(res, expected_result)
+})
+
+test_that("You can perform the window function cume_dist", {
+	  rel_a <- rel_from_df(con, data.frame(a=c(1, 1, 2, 2, 2)))
+    cume_dist_func <- expr_function("cume_dist", list(), order_bys=list(expr_reference("a")))
+    cume_dist_window <- expr_window(cume_dist_func)
+    expr_set_alias(cume_dist_window, "cume_dist")
+    window_proj <- rel_project(rel_a, list(expr_reference("a"), cume_dist_window))
+    order_proj <- rel_order(window_proj, list(expr_reference("a")))
+    res <- rel_to_altrep(order_proj)
+    expected_result <- data.frame(a=c(1, 1, 2, 2, 2), cume_dist=c(0.4, 0.4, 1.0, 1.0, 1.0))
+    expect_equal(res, expected_result)
+})
+
+test_that("You can perform the window function percent rank", {
```

Comment:
Maybe add a test of the `list` aggregate? (It is the only one I know of that doesn't support simple aggregation and it produces lists.)

Reply:
Tried adding a test for this, but couldn't get the list() function to work. I don't think the `list` function is supported at all in the relational api right now, so it might be nice to add support for it in a separate PR

Reply:
Fair enough!